kuksa/edge_client.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) 2024 Contributors to the Eclipse Foundation
+#
+# See the NOTICE file(s) distributed with this work for additional
+# information regarding copyright ownership.
+#
+# This program and the accompanying materials are made available under the
+# terms of the Eclipse Public License 2.0 which is available at
+# https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+# which is available at https://www.apache.org/licenses/LICENSE-2.0.
+#
+# SPDX-License-IDentifier: EPL-2.0 OR Apache-2.0
+
+import argparse
+import signal
+import sys
+
+import paho.mqtt.client as mqtt
+from ditto.client import Client
+from ditto.model.feature import Feature
+from ditto.protocol.things.commands import Command
+from kuksa_client import KuksaClientThread
+
+from edge_device_info import EdgeDeviceInfo
+from utils import process_tree, process_signal
+
+FEATURE_ID_VSS = "VSS"
+
+EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO = "edge/thing/response"
+EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO_REQUEST = "edge/thing/request"
+
+VSS_PATHS = "Vehicle.CurrentLocation.Altitude,Vehicle.CurrentLocation.Latitude,Vehicle.CurrentLocation.Longitude,Vehicle.Speed"
+
+class EdgeClient:
+    def __init__(self, host, port, paths):
+        self.mqtt_client = None
+        self.kuksa_client = KuksaClientThread(config={'ip':host,'protocol': 'grpc', 'port': port, 'insecure': True})
+        self.device_info = None
+        self.ditto_client = None
+        self.vss_paths = paths
+
+    def on_connect(self, client:mqtt.Client, obj, flags, rc):
+        print("Connected with result code:", str(rc))
+        self.mqtt_client = client
+        # init ditto client
+        self.ditto_client = Client(paho_client=self.mqtt_client)
+        self.ditto_client.connect()
+        # init kuksa client
+        self.kuksa_client.start()
+        # trigger initialization
+        self.mqtt_client.subscribe(EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO)
+        self.mqtt_client.publish(EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO_REQUEST, None, 1)
+
+    def on_message(self, client, userdata, msg):
+        try:
+            if msg.topic == EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO:
+                if self.device_info is None:
+                    self.device_info = EdgeDeviceInfo()
+                    self.device_info.unmarshal_json(msg.payload)
+                    self.add_vss_feature()
+                    self.subscribe()
+                else:
+                    print('Device info already available - discarding message')
```

Comment:
the `logging` package would be better to be used instead of `print()`, you can check [this example](https://github.com/k-gostev/robo-demo/blob/main/device/device.py)